src/jit/optimizer.cpp

```diff
@@ -2834,365 +2829,373 @@ void Compiler::optUnrollLoops()
 
-            /* Don't unroll if too much code duplication would result. */
+        /* Don't unroll if too much code duplication would result. */
 
-            if (unrollCostSz > unrollLimitSz)
-            {
-                /* prevent this loop from being revisited */
-                optLoopTable[lnum].lpFlags |= LPFLG_DONT_UNROLL;
-                goto DONE_LOOP;
-            }
+        if (unrollCostSz > unrollLimitSz)
+        {
+            /* prevent this loop from being revisited */
+            optLoopTable[lnum].lpFlags |= LPFLG_DONT_UNROLL;
+            goto DONE_LOOP;
+        }
 
-            /* Looks like a good idea to unroll this loop, let's do it! */
-            CLANG_FORMAT_COMMENT_ANCHOR;
+        /* Looks like a good idea to unroll this loop, let's do it! */
+        CLANG_FORMAT_COMMENT_ANCHOR;
 
 #ifdef DEBUG
-            if (verbose)
+        if (verbose)
+        {
+            printf("\nUnrolling loop BB%02u", head->bbNext->bbNum);
+            if (head->bbNext->bbNum != bottom->bbNum)
             {
-                printf("\nUnrolling loop BB%02u", head->bbNext->bbNum);
-                if (head->bbNext->bbNum != bottom->bbNum)
-                {
-                    printf("..BB%02u", bottom->bbNum);
-                }
-                printf(" over V%02u from %u to %u", lvar, lbeg, llim);
-                printf(" unrollCostSz = %d\n", unrollCostSz);
-                printf("\n");
+                printf("..BB%02u", bottom->bbNum);
             }
+            printf(" over V%02u from %u to %u", lvar, lbeg, llim);
+            printf(" unrollCostSz = %d\n", unrollCostSz);
+            printf("\n");
+        }
 #endif
 
-            /* Create the unrolled loop statement list */
-
-            loopList = loopLast = nullptr;
+        /* Create the unrolled loop statement list */
+        {
+            BlockToBlockMap blockMap(getAllocator());
+            BasicBlock*     insertAfter = bottom;
 
             for (lval = lbeg; totalIter; totalIter--)
             {
-                block = head;
-
-                do
+                for (block = head->bbNext;; block = block->bbNext)
                 {
-                    GenTreeStmt* stmt;
-                    GenTree*     expr;
+                    BasicBlock* newBlock = insertAfter =
+                        fgNewBBafter(block->bbJumpKind, insertAfter, /*extendRegion*/ true);
+                    blockMap.Set(block, newBlock);
 
-                    block = block->bbNext;
-                    noway_assert(block);
-
-                    /* Visit all the statements in the block */
-
-                    for (stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
+                    if (!BasicBlock::CloneBlockState(this, newBlock, block, lvar, lval))
                     {
-                        /* Stop if we've reached the end of the loop */
+                        // cloneExpr doesn't handle everything
```

Comment:
Is there any intermediate information left over that we need to restore? I'm thinking in particular about lclVar ref counts.